File: projects/lang-repl/parser.py
"""
Parser for the LangRePL programming language.
Builds an Abstract Syntax Tree (AST) from tokens.
"""

from typing import List, Optional, Union
from lexer import Token, TokenType


# AST Node Types
class Expr:
    """Base class for expression nodes."""
    pass


class Binary(Expr):
    def __init__(self, left: Expr, operator: Token, right: Expr):
        self.left = left
        self.operator = operator
        self.right = right
    
    def __repr__(self):
        return f"Binary({self.left}, {self.operator.lexeme}, {self.right})"


class Unary(Expr):
    def __init__(self, operator: Token, right: Expr):
        self.operator = operator
        self.right = right
    
    def __repr__(self):
        return f"Unary({self.operator.lexeme}, {self.right})"


class Literal(Expr):
    def __init__(self, value: object):
        self.value = value
    
    def __repr__(self):
        return f"Literal({self.value})"


class Grouping(Expr):
    def __init__(self, expression: Expr):
        self.expression = expression
    
    def __repr__(self):
        return f"Grouping({self.expression})"


class Variable(Expr):
    def __init__(self, name: Token):
        self.name = name
    
    def __repr__(self):
        return f"Variable({self.name.lexeme})"


class Assignment(Expr):
    def __init__(self, name: Token, value: Expr):
        self.name = name
        self.value = value
    
    def __repr__(self):
        return f"Assignment({self.name.lexeme}, {self.value})"


class Call(Expr):
    def __init__(self, callee: Expr, paren: Token, arguments: List[Expr]):
        self.callee = callee
        self.paren = paren
        self.arguments = arguments
    
    def __repr__(self):
        return f"Call({self.callee}, {self.arguments})"


class Get(Expr):
    def __init__(self, object: Expr, name: Token):
        self.object = object
        self.name = name
    
    def __repr__(self):
        return f"Get({self.object}, {self.name.lexeme})"


class Array(Expr):
    def __init__(self, elements: List[Expr]):
        self.elements = elements
    
    def __repr__(self):
        return f"Array({self.elements})"


class Index(Expr):
    def __init__(self, array: Expr, index: Expr):
        self.array = array
        self.index = index
    
    def __repr__(self):
        return f"Index({self.array}, {self.index})"


# Statement types
class Stmt:
    """Base class for statement nodes."""
    pass


class ExpressionStmt(Stmt):
    def __init__(self, expression: Expr):
        self.expression = expression
    
    def __repr__(self):
        return f"ExpressionStmt({self.expression})"


class PrintStmt(Stmt):
    def __init__(self, expression: Expr):
        self.expression = expression
    
    def __repr__(self):
        return f"PrintStmt({self.expression})"


class VarStmt(Stmt):
    def __init__(self, name: Token, initializer: Optional[Expr]):
        self.name = name
        self.initializer = initializer
    
    def __repr__(self):
        return f"VarStmt({self.name.lexeme}, {self.initializer})"


class BlockStmt(Stmt):
    def __init__(self, statements: List[Stmt]):
        self.statements = statements
    
    def __repr__(self):
        return f"BlockStmt({self.statements})"


class IfStmt(Stmt):
    def __init__(self, condition: Expr, then_branch: Stmt, else_branch: Optional[Stmt]):
        self.condition = condition
        self.then_branch = then_branch
        self.else_branch = else_branch
    
    def __repr__(self):
        return f"IfStmt({self.condition}, {self.then_branch}, {self.else_branch})"


class WhileStmt(Stmt):
    def __init__(self, condition: Expr, body: Stmt):
        self.condition = condition
        self.body = body
    
    def __repr__(self):
        return f"WhileStmt({self.condition}, {self.body})"


class ForStmt(Stmt):
    def __init__(self, initializer: Optional[Stmt], condition: Optional[Expr], 
                 increment: Optional[Expr], body: Stmt):
        self.initializer = initializer
        self.condition = condition
        self.increment = increment
        self.body = body
    
    def __repr__(self):
        return f"ForStmt({self.initializer}, {self.condition}, {self.increment}, {self.body})"


class FunctionStmt(Stmt):
    def __init__(self, name: Token, params: List[Token], body: List[Stmt]):
        self.name = name
        self.params = params
        self.body = body
    
    def __repr__(self):
        return f"FunctionStmt({self.name.lexeme}, {[p.lexeme for p in self.params]}, {self.body})"


class ReturnStmt(Stmt):
    def __init__(self, keyword: Token, value: Optional[Expr]):
        self.keyword = keyword
        self.value = value
    
    def __repr__(self):
        return f"ReturnStmt({self.value})"


class Parser:
    """Recursive descent parser for LangRePL."""
    
    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.current = 0
    
    def parse(self) -> List[Stmt]:
        """Parse the entire token list into statements."""
        statements = []
        
        while not self._at_end():
            if self._check(TokenType.NEWLINE):
                self._advance()
                continue
            
            stmt = self._declaration()
            if stmt:
                statements.append(stmt)
        
        return statements
    
    def _at_end(self) -> bool:
        return self._peek().type == TokenType.EOF
    
    def _peek(self) -> Token:
        return self.tokens[self.current]
    
    def _previous(self) -> Token:
        return self.tokens[self.current - 1]
    
    def _advance(self) -> Token:
        if not self._at_end():
            self.current += 1
        return self._previous()
    
    def _check(self, token_type: TokenType) -> bool:
        if self._at_end():
            return False
        return self._peek().type == token_type
    
    def _match(self, *token_types: TokenType) -> bool:
        for token_type in token_types:
            if self._check(token_type):
                self._advance()
                return True
        return False
    
    def _consume(self, token_type: TokenType, message: str) -> Token:
        if self._check(token_type):
            return self._advance()
        
        raise SyntaxError(f"{message} at line {self._peek().line}")
    
    def _synchronize(self):
        """Synchronize parser after an error."""
        self._advance()
        
        while not self._at_end():
            if self._previous().type == TokenType.SEMICOLON:
                return
            
            if self._peek().type in (TokenType.CLASS, TokenType.FN, TokenType.LET, 
                                     TokenType.IF, TokenType.WHILE, TokenType.FOR,
                                     TokenType.PRINT, TokenType.RETURN):
                return
            
            self._advance()
    
    # Declaration parsing
    def _declaration(self) -> Optional[Stmt]:
        try:
            if self._match(TokenType.FN):
                return self._function("function")
            if self._match(TokenType.LET):
                return self._var_declaration()
            
            return self._statement()
        except SyntaxError as e:
            self._synchronize()
            raise e
    
    def _function(self, kind: str) -> FunctionStmt:
        name = self._consume(TokenType.IDENTIFIER, f"Expect {kind} name")
        
        self._consume(TokenType.LPAREN, f"Expect '(' after {kind} name")
        params = []
        
        if not self._check(TokenType.RPAREN):
            params.append(self._consume(TokenType.IDENTIFIER, "Expect parameter name"))
            while self._match(TokenType.COMMA):
                params.append(self._consume(TokenType.IDENTIFIER, "Expect parameter name"))
        
        self._consume(TokenType.RPAREN, "Expect ')' after parameters")
        self._consume(TokenType.LBRACE, f"Expect '{{' before {kind} body")
        
        body = self._block()
        return FunctionStmt(name, params, body)
    
    def _var_declaration(self) -> VarStmt:
        name = self._consume(TokenType.IDENTIFIER, "Expect variable name")
        
        initializer = None
        if self._match(TokenType.ASSIGN):
            initializer = self._expression()
        
        self._consume(TokenType.SEMICOLON, "Expect ';' after variable declaration")
        return VarStmt(name, initializer)
    
    # Statement parsing
    def _statement(self) -> Stmt:
        if self._match(TokenType.IF):
            return self._if_statement()
        if self._match(TokenType.WHILE):
            return self._while_statement()
        if self._match(TokenType.FOR):
            return self._for_statement()
        if self._match(TokenType.PRINT):
            return self._print_statement()
        if self._match(TokenType.RETURN):
            return self._return_statement()
        if self._match(TokenType.LBRACE):
            return BlockStmt(self._block())
        
        return self._expression_statement()
    
    def _if_statement(self) -> IfStmt:
        self._consume(TokenType.LPAREN, "Expect '(' after 'if'")
        condition = self._expression()
        self._consume(TokenType.RPAREN, "Expect ')' after if condition")
        
        then_branch = self._statement()
        else_branch = None
        
        if self._match(TokenType.ELSE):
            else_branch = self._statement()
        
        return IfStmt(condition, then_branch, else_branch)
    
    def _while_statement(self) -> WhileStmt:
        self._consume(TokenType.LPAREN, "Expect '(' after 'while'")
        condition = self._expression()
        self._consume(TokenType.RPAREN, "Expect ')' after while condition")
        
        body = self._statement()
        return WhileStmt(condition, body)
    
    def _for_statement(self) -> ForStmt:
        self._consume(TokenType.LPAREN, "Expect '(' after 'for'")
        
        initializer = None
        if self._match(TokenType.SEMICOLON):
            pass
        elif self._match(TokenType.LET):
            initializer = self._var_declaration()
        else:
            initializer = self._expression_statement()
        
        condition = None
        if not self._check(TokenType.SEMICOLON):
            condition = self._expression()
        self._consume(TokenType.SEMICOLON, "Expect ';' after loop condition")
        
        increment = None
        if not self._check(TokenType.RPAREN):
            increment = self._expression()
        self._consume(TokenType.RPAREN, "Expect ')' after for clauses")
        
        body = self._statement()
        
        # Desugar for loop to while loop
        # for (init; cond; incr) body -> { init; while (cond) { body; incr; } }
        if increment:
            body = BlockStmt([body, ExpressionStmt(increment)])
        
        if condition is None:
            condition = Literal(True)
        
        body = WhileStmt(condition, body)
        
        if initializer:
            body = BlockStmt([initializer, body])
        
        return body
    
    def _print_statement(self) -> PrintStmt:
        value = self._expression()
        self._consume(TokenType.SEMICOLON, "Expect ';' after value")
        return PrintStmt(value)
    
    def _return_statement(self) -> ReturnStmt:
        keyword = self._previous()
        value = None
        
        if not self._check(TokenType.SEMICOLON):
            value = self._expression()
        
        self._consume(TokenType.SEMICOLON, "Expect ';' after return value")
        return ReturnStmt(keyword, value)
    
    def _block(self) -> List[Stmt]:
        statements = []
        
        while not self._check(TokenType.RBRACE) and not self._at_end():
            if self._check(TokenType.NEWLINE):
                self._advance()
                continue
            
            stmt = self._declaration()
            if stmt:
                statements.append(stmt)
        
        self._consume(TokenType.RBRACE, "Expect '}' after block")
        return statements
    
    def _expression_statement(self) -> ExpressionStmt:
        expr = self._expression()
        self._consume(TokenType.SEMICOLON, "Expect ';' after expression")
        return ExpressionStmt(expr)
    
    # Expression parsing (precedence climbing)
    def _expression(self) -> Expr:
        return self._assignment()
    
    def _assignment(self) -> Expr:
        expr = self._or()
        
        if self._match(TokenType.ASSIGN, TokenType.PLUS_ASSIGN, TokenType.MINUS_ASSIGN):
            operator = self._previous()
            value = self._assignment()
            
            if isinstance(expr, Variable):
                if operator.type == TokenType.PLUS_ASSIGN:
                    value = Binary(expr, Token(TokenType.PLUS, "+", None, operator.line), value)
                elif operator.type == TokenType.MINUS_ASSIGN:
                    value = Binary(expr, Token(TokenType.MINUS, "-", None, operator.line), value)
                return Assignment(expr.name, value)
            
            raise SyntaxError(f"Invalid assignment target at line {operator.line}")
        
        return expr
    
    def _or(self) -> Expr:
        expr = self._and()
        
        while self._match(TokenType.OR):
            operator = self._previous()
            right = self._and()
            expr = Binary(expr, operator, right)
        
        return expr
    
    def _and(self) -> Expr:
        expr = self._equality()
        
        while self._match(TokenType.AND):
            operator = self._previous()
            right = self._equality()
            expr = Binary(expr, operator, right)
        
        return expr
    
    def _equality(self) -> Expr:
        expr = self._comparison()
        
        while self._match(TokenType.BANG_EQUAL, TokenType.EQUAL_EQUAL):
            operator = self._previous()
            right = self._comparison()
            expr = Binary(expr, operator, right)
        
        return expr
    
    def _comparison(self) -> Expr:
        expr = self._term()
        
        while self._match(TokenType.GREATER, TokenType.GREATER_EQUAL, 
                          TokenType.LESS, TokenType.LESS_EQUAL):
            operator = self._previous()
            right = self._term()
            expr = Binary(expr, operator, right)
        
        return expr
    
    def _term(self) -> Expr:
        expr = self._factor()
        
        while self._match(TokenType.MINUS, TokenType.PLUS):
            operator = self._previous()
            right = self._factor()
            expr = Binary(expr, operator, right)
        
        return expr
    
    def _factor(self) -> Expr:
        expr = self._unary()
        
        while self._match(TokenType.SLASH, TokenType.STAR, TokenType.PERCENT):
            operator = self._previous()
            right = self._unary()
            expr = Binary(expr, operator, right)
        
        return expr
    
    def _unary(self) -> Expr:
        if self._match(TokenType.BANG, TokenType.MINUS):
            operator = self._previous()
            right = self._unary()
            return Unary(operator, right)
        
        return self._call()
    
    def _call(self) -> Expr:
        expr = self._primary()
        
        while True:
            if self._match(TokenType.LPAREN):
                expr = self._finish_call(expr)
            elif self._match(TokenType.LBRACKET):
                index = self._expression()
                self._consume(TokenType.RBRACKET, "Expect ']' after index")
                expr = Index(expr, index)
            else:
                break
        
        return expr
    
    def _finish_call(self, callee: Expr) -> Expr:
        arguments = []
        
        if not self._check(TokenType.RPAREN):
            arguments.append(self._expression())
            while self._match(TokenType.COMMA):
                arguments.append(self._expression())
        
        self._consume(TokenType.RPAREN, "Expect ')' after arguments")
        return Call(callee, self._previous(), arguments)
    
    def _primary(self) -> Expr:
        if self._match(TokenType.TRUE):
            return Literal(True)
        if self._match(TokenType.FALSE):
            return Literal(False)
        if self._match(TokenType.NIL):
            return Literal(None)
        
        if self._match(TokenType.NUMBER):
            return Literal(self._previous().literal)
        
        if self._match(TokenType.STRING):
            return Literal(self._previous().literal)
        
        if self._match(TokenType.LBRACKET):
            return self._array()
        
        if self._match(TokenType.LPAREN):
            expr = self._expression()
            self._consume(TokenType.RPAREN, "Expect ')' after expression")
            return Grouping(expr)
        
        if self._match(TokenType.IDENTIFIER):
            return Variable(self._previous())
        
        raise SyntaxError(f"Expect expression at line {self._peek().line}")
    
    def _array(self) -> Expr:
        elements = []
        
        if not self._check(TokenType.RBRACKET):
            elements.append(self._expression())
            while self._match(TokenType.COMMA):
                elements.append(self._expression())
        
        self._consume(TokenType.RBRACKET, "Expect ']' after array elements")
        return Array(elements)

File: projects/micro-compiler/src/parser/Parser.ts
import { Token, TokenType } from '../types/Token';
import {
  ASTNode,
  Program,
  Statement,
  Expression,
  ExpressionStatement,
  VariableDeclaration,
  BlockStatement,
  IfStatement,
  WhileStatement,
  FunctionDeclaration,
  PrintStatement,
  BinaryExpression,
  UnaryExpression,
  LiteralExpression,
  IdentifierExpression,
  CallExpression,
  AssignmentExpression
} from '../types/AST';
import { createValue } from '../types/Value';

/**
 * Parser for the Volt language
 * Builds an Abstract Syntax Tree from tokens
 */
export class Parser {
  private tokens: Token[];
  private current: number = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  /**
   * Parse the entire token list into an AST
   */
  public parse(): Program {
    const statements: Statement[] = [];

    while (!this.isAtEnd()) {
      const stmt = this.declaration();
      if (stmt) {
        statements.push(stmt);
      }
    }

    return { type: 'Program', statements };
  }

  /**
   * Parse a declaration (variable or function)
   */
  private declaration(): Statement | null {
    try {
      if (this.match(TokenType.FUNCTION)) {
        return this.functionDeclaration();
      }
      if (this.match(TokenType.LET)) {
        return this.variableDeclaration();
      }
      return this.statement();
    } catch (error) {
      this.synchronize();
      return null;
    }
  }

  /**
   * Parse a function declaration
   */
  private functionDeclaration(): FunctionDeclaration {
    const name = this.consume(TokenType.IDENTIFIER, 'Expect function name').lexeme;

    this.consume(TokenType.LEFT_PAREN, "Expect '(' after function name");
    const parameters: string[] = [];

    if (!this.check(TokenType.RIGHT_PAREN)) {
      do {
        const param = this.consume(TokenType.IDENTIFIER, 'Expect parameter name').lexeme;
        parameters.push(param);
      } while (this.match(TokenType.COMMA));
    }

    this.consume(TokenType.RIGHT_PAREN, "Expect ')' after parameters");
    this.consume(TokenType.LEFT_BRACE, "Expect '{' before function body");
    const body = this.block();

    return {
      type: 'FunctionDeclaration',
      name,
      parameters,
      body
    };
  }

  /**
   * Parse a variable declaration
   */
  private variableDeclaration(): VariableDeclaration {
    const name = this.consume(TokenType.IDENTIFIER, 'Expect variable name').lexeme;

    let initializer: Expression | undefined;
    if (this.match(TokenType.EQUAL)) {
      initializer = this.expression();
    }

    this.consume(TokenType.SEMICOLON, "Expect ';' after variable declaration");

    return {
      type: 'VariableDeclaration',
      name,
      initializer
    };
  }

  /**
   * Parse a statement
   */
  private statement(): Statement {
    if (this.match(TokenType.IF)) {
      return this.ifStatement();
    }
    if (this.match(TokenType.WHILE)) {
      return this.whileStatement();
    }
    if (this.match(TokenType.LEFT_BRACE)) {
      return this.block();
    }
    if (this.match(TokenType.PRINT)) {
      return this.printStatement();
    }

    return this.expressionStatement();
  }

  /**
   * Parse an if statement
   */
  private ifStatement(): IfStatement {
    this.consume(TokenType.LEFT_PAREN, "Expect '(' after 'if'");
    const condition = this.expression();
    this.consume(TokenType.RIGHT_PAREN, "Expect ')' after if condition");

    const thenBranch = this.statement();
    let elseBranch: BlockStatement | null = null;

    if (this.match(TokenType.ELSE)) {
      elseBranch = this.statement() as BlockStatement;
    }

    return {
      type: 'IfStatement',
      condition,
      thenBranch,
      elseBranch
    };
  }

  /**
   * Parse a while statement
   */
  private whileStatement(): WhileStatement {
    this.consume(TokenType.LEFT_PAREN, "Expect '(' after 'while'");
    const condition = this.expression();
    this.consume(TokenType.RIGHT_PAREN, "Expect ')' after while condition");

    const body = this.statement();

    return {
      type: 'WhileStatement',
      condition,
      body
    };
  }

  /**
   * Parse a block statement
   */
  private block(): BlockStatement {
    const statements: Statement[] = [];

    while (!this.check(TokenType.RIGHT_BRACE) && !this.isAtEnd()) {
      const stmt = this.declaration();
      if (stmt) {
        statements.push(stmt);
      }
    }

    this.consume(TokenType.RIGHT_BRACE, "Expect '}' after block");

    return {
      type: 'BlockStatement',
      statements
    };
  }

  /**
   * Parse a print statement
   */
  private printStatement(): PrintStatement {
    this.consume(TokenType.LEFT_PAREN, "Expect '(' after 'print'");
    const expression = this.expression();
    this.consume(TokenType.RIGHT_PAREN, "Expect ')' after print expression");
    this.consume(TokenType.SEMICOLON, "Expect ';' after print statement");

    return {
      type: 'PrintStatement',
      expression
    };
  }

  /**
   * Parse an expression statement
   */
  private expressionStatement(): ExpressionStatement {
    const expr = this.expression();
    this.consume(TokenType.SEMICOLON, "Expect ';' after expression");
    return {
      type: 'ExpressionStatement',
      expression: expr
    };
  }

  /**
   * Parse an expression
   */
  private expression(): Expression {
    return this.assignment();
  }

  /**
   * Parse an assignment expression
   */
  private assignment(): Expression {
    const expr = this.equality();

    if (this.match(TokenType.EQUAL)) {
      const equals = this.previous();
      const value = this.assignment();

      if (expr.type === 'IdentifierExpression') {
        return {
          type: 'AssignmentExpression',
          name: (expr as IdentifierExpression).name,
          value
        } as AssignmentExpression;
      }

      throw new Error('Invalid assignment target');
    }

    return expr;
  }

  /**
   * Parse equality expressions (==, !=)
   */
  private equality(): Expression {
    let expr = this.comparison();

    while (this.match(TokenType.EQUAL_EQUAL, TokenType.BANG_EQUAL)) {
      const operator = this.previous().lexeme;
      const right = this.comparison();
      expr = {
        type: 'BinaryExpression',
        operator,
        left: expr,
        right
      } as BinaryExpression;
    }

    return expr;
  }

  /**
   * Parse comparison expressions (<, >, <=, >=)
   */
  private comparison(): Expression {
    let expr = this.term();

    while (
      this.match(
        TokenType.LESS,
        TokenType.LESS_EQUAL,
        TokenType.GREATER,
        TokenType.GREATER_EQUAL
      )
    ) {
      const operator = this.previous().lexeme;
      const right = this.term();
      expr = {
        type: 'BinaryExpression',
        operator,
        left: expr,
        right
      } as BinaryExpression;
    }

    return expr;
  }

  /**
   * Parse term expressions (+, -)
   */
  private term(): Expression {
    let expr = this.factor();

    while (this.match(TokenType.PLUS, TokenType.MINUS)) {
      const operator = this.previous().lexeme;
      const right = this.factor();
      expr = {
        type: 'BinaryExpression',
        operator,
        left: expr,
        right
      } as BinaryExpression;
    }

    return expr;
  }

  /**
   * Parse factor expressions (*, /, %)
   */
  private factor(): Expression {
    let expr = this.unary();

    while (this.match(TokenType.STAR, TokenType.SLASH, TokenType.PERCENT)) {
      const operator = this.previous().lexeme;
      const right = this.unary();
      expr = {
        type: 'BinaryExpression',
        operator,
        left: expr,
        right
      } as BinaryExpression;
    }

    return expr;
  }

  /**
   * Parse unary expressions (-)
   */
  private unary(): Expression {
    if (this.match(TokenType.MINUS)) {
      const operator = this.previous().lexeme;
      const right = this.unary();
      return {
        type: 'UnaryExpression',
        operator,
        operand: right
      } as UnaryExpression;
    }

    return this.primary();
  }

  /**
   * Parse primary expressions
   */
  private primary(): Expression {
    if (this.match(TokenType.NUMBER)) {
      return {
        type: 'LiteralExpression',
        value: createValue(this.previous().literal)
      } as LiteralExpression;
    }

    if (this.match(TokenType.STRING)) {
      return {
        type: 'LiteralExpression',
        value: createValue(this.previous().literal)
      } as LiteralExpression;
    }

    if (this.match(TokenType.TRUE)) {
      return {
        type: 'LiteralExpression',
        value: createValue(true)
      } as LiteralExpression;
    }

    if (this.match(TokenType.FALSE)) {
      return {
        type: 'LiteralExpression',
        value: createValue(false)
      } as LiteralExpression;
    }

    if (this.match(TokenType.IDENTIFIER)) {
      // Save the identifier token before checking for function call
      const identifier = this.previous();
      
      // Check if this is a function call
      if (this.match(TokenType.LEFT_PAREN)) {
        return this.finishCall(identifier);
      }

      return {
        type: 'IdentifierExpression',
        name: identifier.lexeme
      } as IdentifierExpression;
    }

    if (this.match(TokenType.LEFT_PAREN)) {
      const expr = this.expression();
      this.consume(TokenType.RIGHT_PAREN, "Expect ')' after expression");
      return expr;
    }

    throw new Error(`Expect expression at line ${this.peek().line}`);
  }

  /**
   * Finish parsing a function call
   */
  private finishCall(callee: Token): CallExpression {
    const args: Expression[] = [];

    if (!this.check(TokenType.RIGHT_PAREN)) {
      do {
        args.push(this.expression());
      } while (this.match(TokenType.COMMA));
    }

    this.consume(TokenType.RIGHT_PAREN, "Expect ')' after arguments");

    return {
      type: 'CallExpression',
      callee: {
        type: 'IdentifierExpression',
        name: callee.lexeme
      } as IdentifierExpression,
      arguments: args
    } as CallExpression;
  }

  /**
   * Check if current token matches any of the given types
   */
  private match(...types: TokenType[]): boolean {
    for (const type of types) {
      if (this.check(type)) {
        this.advance();
        return true;
      }
    }
    return false;
  }

  /**
   * Check if current token is of given type
   */
  private check(type: TokenType): boolean {
    if (this.isAtEnd()) {
      return false;
    }
    return this.peek().type === type;
  }

  /**
   * Consume current token if it matches type, otherwise throw error
   */
  private consume(type: TokenType, message: string): Token {
    if (this.check(type)) {
      return this.advance();
    }
    throw new Error(`${message} at line ${this.peek().line}`);
  }

  /**
   * Advance to next token and return previous
   */
  private advance(): Token {
    if (!this.isAtEnd()) {
      this.current++;
    }
    return this.tokens[this.current - 1];
  }

  /**
   * Check if at end of tokens
   */
  private isAtEnd(): boolean {
    return this.peek().type === TokenType.EOF;
  }

  /**
   * Get current token
   */
  private peek(): Token {
    return this.tokens[this.current];
  }

  /**
   * Get previous token
   */
  private previous(): Token {
    return this.tokens[this.current - 1];
  }

  /**
   * Synchronize parser after error
   */
  private synchronize(): void {
    this.advance();

    while (!this.isAtEnd()) {
      if (this.previous().type === TokenType.SEMICOLON) {
        return;
      }

      switch (this.peek().type) {
        case TokenType.FUNCTION:
        case TokenType.LET:
        case TokenType.IF:
        case TokenType.WHILE:
        case TokenType.PRINT:
          return;
      }

      this.advance();
    }
  }
}